cdc/server.go

```diff
@@ -350,3 +360,110 @@ func (s *Server) Close() {
 		s.statusServer = nil
 	}
 }
+
+func (s *Server) setUpDataDir(ctx context.Context) error {
+	conf := config.GetGlobalServerConfig()
+	if conf.DataDir != "" {
+		conf.Sorter.SortDir = filepath.Join(conf.DataDir, config.DefaultSortDir)
+		config.StoreGlobalServerConfig(conf)
+
+		return nil
+	}
+
+	// data-dir will be decide by exist changefeed for backward compatibility
+	allStatus, err := s.etcdClient.GetAllChangeFeedStatus(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	candidates := make([]string, 0, len(allStatus))
+	for id := range allStatus {
+		info, err := s.etcdClient.GetChangeFeedInfo(ctx, id)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if info.SortDir != "" {
+			candidates = append(candidates, info.SortDir)
+		}
+	}
+
+	conf.DataDir = defaultDataDir
+	best, ok := findBestDataDir(candidates)
+	if ok {
+		conf.DataDir = best
+	}
+
+	conf.Sorter.SortDir = filepath.Join(conf.DataDir, config.DefaultSortDir)
+	config.StoreGlobalServerConfig(conf)
+	return nil
+}
+
+func (s *Server) initDataDir(ctx context.Context) error {
+	if err := s.setUpDataDir(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	conf := config.GetGlobalServerConfig()
+	err := os.MkdirAll(conf.DataDir, 0o755)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	diskInfo, err := util.GetDiskInfo(conf.DataDir)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if diskInfo.Avail < dataDirThreshold {
+		log.Warn(fmt.Sprintf("%s is set as data-dir (%dGB available), ticdc recommend disk for data-dir "+
+			"at least have %dGB available space", conf.DataDir, diskInfo.Avail, dataDirThreshold))
+	}
+
+	return nil
+}
+
+// try to find the best data dir by rules
+// at the moment, only consider available disk space
+func findBestDataDir(dirs []string) (result string, ok bool) {
+	var (
+		low        uint64 = 0
+		candidates        = getDataDirCandidates(dirs)
+	)
+
+	for dir := range candidates {
+		if err := util.IsDirReadWritable(dir); err != nil {
+			log.Warn("try to get disk info failed", zap.String("dir", dir), zap.Error(err))
+			continue
+		}
+		info, err := util.GetDiskInfo(dir)
+		if err != nil {
+			log.Warn("try to get disk info failed", zap.String("dir", dir), zap.Error(err))
+			continue
+		}
+		if info.Avail > low {
+			result = dir
+			low = info.Avail
+			ok = true
+		}
+	}
+
+	if !ok && len(dirs) != 0 {
+		log.Warn("try to find directory for data-dir failed, use `/tmp/cdc_data` as data-dir", zap.Strings("candidates", dirs))
+	}
+
+	return result, ok
+}
+
+func getDataDirCandidates(dirs []string) (result map[string]struct{}) {
```

Comment:
Does this function try to find a `data-dir` based on given `sort-dir`?

Reply:
Use the given `sort-dir`directly as `data-dir` for server is not a good choice. Consider `/tmp/cdc_sort` was an old changefeed `sort-dir`, it's not good as server `data-dir`

Reply:
IMO the reason we should use one of the old `sort-dir` is to keep the changefeed can run normally after upgrading.

Considering user has set `sort-dir` to `/data/ticdc/sort-dir`, which is mounted with a 1T SSD, the `/` path could only has 100GB, which is used to store binaries and libraries. In this scenario using `/` as data-dir is not a good choice.

Reply:
I got your point, your idea is more considerable. We have to mention user how to start server with data-dir and the logic behind data-dir chosen.